--- a/ external/zxcvbn-c/UnpackedTarball_zxcvbn-c.mk
+++ b/external/zxcvbn-c/UnpackedTarball_zxcvbn-c.mk
@@ -11,8 +11,9 @@ $(eval $(call gb_UnpackedTarball_Unpacke
 
 $(eval $(call gb_UnpackedTarball_set_tarball,zxcvbn-c,$(ZXCVBN_C_TARBALL)))
 
 $(eval $(call gb_UnpackedTarball_add_patches,zxcvbn-c,\
	external/zxcvbn-c/zxcvbn-c-2.5-do-not-use-stdafx.patch \
+	external/zxcvbn-c/0001-There-is-no-std-basic_string-int.patch \
 ))
 
 # vim: set noet sw=4 ts=4:
