diff --git a/node.gyp b/node.gyp
index 8fe89d8b1bc..12204cb2f60 100644
--- a/node.gyp
+++ b/node.gyp
@@ -115,7 +115,9 @@
           '-Wl,-bnoerrmsg',
         ],
       }],
-      ['OS == "linux" and llvm_version != "0.0"', {
+      ['OS == "linux" and '
+       'llvm_version != "0.0" and '
+       'target_arch in "arm loong64 mips mips64 mips64el mipsel ppc riscv64"', {
         'libraries': ['-latomic'],
       }],
     ],
