--- a/configure.ac	2021-12-01 21:12:17.943229767 -0500
+++ b/configure.ac	2021-12-01 21:14:19.409602639 -0500
@@ -1338,11 +1338,12 @@ if test "x$enable_stack_protector" = xno
     # Need that, because some distributions ship compilers that include
     # `-fstack-protector' in the default specs.
     TARGET_CFLAGS="$TARGET_CFLAGS -fno-stack-protector"
   fi
 elif test "x$platform" != xefi; then
-  AC_MSG_ERROR([--enable-stack-protector is only supported on EFI platforms])
+  AC_MSG_WARN([--enable-stack-protector is only supported on EFI platforms])
+  enable_stack_protector=no
 elif test "x$ssp_global_possible" != xyes; then
   AC_MSG_ERROR([--enable-stack-protector is not supported (compiler doesn't support -mstack-protector-guard=global)])
 else
   TARGET_CFLAGS="$TARGET_CFLAGS -mstack-protector-guard=global"
   if test "x$enable_stack_protector" = xyes; then
